src/utils/include/utils/LCA.hpp

```diff
@@ -15,100 +13,101 @@ class LCA
 public:
     LCA() = default;
 
-    void        addEdge( std::string father, std::string son );
+    void        addEdge( const std::string& father, const std::string& son );
     void        doEulerWalk();
-    int         getLCA( int u, int v );
-    std::string getLCA( std::vector< std::string >& taxIds );
+    int         getLCA( unsigned int u, unsigned int v ) const;
+    std::string getLCA( const std::vector< std::string >& taxIds ) const;
 
 private:
-    void depthFirstSearch( std::string current, int depth );
+    void depthFirstSearch( const std::string& current, unsigned int depth );
     void preProcessRMQ();
-    int  queryRMQ( int i, int j );
+    int  queryRMQ( unsigned int i, unsigned int j ) const;
+
+    static constexpr int first_appearance_init = -1;
 
     std::unordered_map< std::string, std::vector< std::string > > m_parents;
     std::vector< int >                                            m_euler;
     std::vector< int >                                            m_depth;
-    std::unique_ptr< int[] >                                      m_firstAppearance;
+    std::vector< int >                                            m_firstAppearance;
     int                                                           m_vertices = 0;
-    std::unordered_map< std::string, int >                        m_encode;
-    std::unordered_map< int, std::string >                        m_decode;
-    std::unique_ptr< std::unique_ptr< int[] >[] >                 m_M;
+    std::unordered_map< std::string, unsigned int >               m_encode;
+    std::vector< std::string >                                    m_decode;
+    std::vector< std::vector< int > >                             m_M;
 };
 
-inline void LCA::addEdge( std::string father, std::string son )
+inline void LCA::addEdge( const std::string& father, const std::string& son )
 {
     if ( m_encode.count( father ) == 0 )
     {
         m_encode.insert( { father, m_vertices } );
-        m_decode.insert( { m_vertices, father } );
-        m_vertices++;
+        m_decode.insert( m_decode.begin() + m_vertices, father );
```

Comment:
Crazy! Cool! Let's keep an eye on this... here we allocate quite a lot of strings, but still the outcome when reading could be good!